src/state/assets/utils.ts

```diff
@@ -0,0 +1,471 @@
+  ZerionAsset,
+} from '@/__swaps__/types/assets';
+import { AddressAssetsReceivedMessage } from '@/__swaps__/types/refraction';
+import { parseUserAsset } from '@/__swaps__/utils/assets';
+import { time } from '@/utils';
+import { UserAssetsState, UserAssetsParams } from './types';
+import { userAssetsStore, useUserAssetsStore } from './userAssets';
+import { userAssetsStoreManager } from './userAssetsStoreManager';
+
+const USER_ASSETS_TIMEOUT_DURATION = time.seconds(10);
+
+// ============ Fetch Utils ==================================================== //
+
+export async function fetchUserAssets(
+  params: UserAssetsParams,
+  abortController: AbortController | null
+): Promise<{ chainIdsWithErrors: ChainId[] | null; parsedAssetsDict: ParsedAssetsDictByChain | null } | null> {
+  const { address, currency, testnetMode } = params;
+
+  if (testnetMode) {
+    const { assets, chainIdsInResponse } = await fetchAnvilBalancesByChainId(address);
+    const parsedAssets: Array<{
+      asset: ZerionAsset;
+      quantity: string;
+      small_balances: boolean;
+    }> = Object.values(assets).map(asset => ({
+      asset: asset.asset,
+      quantity: asset.quantity,
+      small_balances: false,
+    }));
+
+    const parsedAssetsDict = parseUserAssets({
+      assets: parsedAssets,
+      chainIds: chainIdsInResponse,
+      currency,
+    });
+
+    return { chainIdsWithErrors: null, parsedAssetsDict };
+  }
+
+  try {
+    staleBalancesStore.getState().clearExpiredData(address);
+    const staleBalanceParam = staleBalancesStore.getState().getStaleBalancesQueryParam(address);
+    let url = `/${useBackendNetworksStore.getState().getSupportedChainIds().join(',')}/${address}/assets?currency=${currency.toLowerCase()}`;
+    if (staleBalanceParam) {
+      url += staleBalanceParam;
+    }
+    const res = await addysHttp.get<AddressAssetsReceivedMessage>(url, {
+      abortController,
+      timeout: USER_ASSETS_TIMEOUT_DURATION,
+    });
+    const chainIdsInResponse = res?.data?.meta?.chain_ids || [];
+    const chainIdsWithErrors = res?.data?.meta?.chain_ids_with_errors || [];
+    const assets = res?.data?.payload?.assets?.filter(asset => !asset.asset.defi_position) || [];
+    if (address) {
+      let parsedAssetsDict: ParsedAssetsDictByChain | null = null;
+
+      if (assets.length && chainIdsInResponse.length) {
+        parsedAssetsDict = parseUserAssets({
+          assets,
+          chainIds: chainIdsInResponse,
+          currency,
+        });
+      } else if (chainIdsWithErrors.length) {
+        /**
+         * Currently only retrying if no asset data is returned, as the chains that individually fail
+         * (e.g. due to rate limiting) are highly likely to fail again if retried immediately. So instead
+         * we simply wait for the next fetch and ensure we don't override existing data for failed chains.
+         */
+        refetchAssetsForFailedChains({
+          address,
+          chainIds: chainIdsWithErrors,
```

Comment:
if we reach this block, `chainIdsWithErrors` will be every chain id included in the original request, right?

Reply:
Yeah I think I saw this happen once in all of my testing, figured it's worth retrying if all fail.